File: practice/weather_ws/src/weather_monitor_py/weather_monitor_py/client_node.py
import rclpy
from rclpy.node import Node
from interfaces.srv import SetThreshold
import sys

class ThresholdClient(Node):
    def __init__(self):
        super().__init__('threshold_client_py')
        # 创建客户端
        self.client = self.create_client(SetThreshold,'set_weather_threshold')

        # 等待服务可用
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('服务不可用，正在等待...')

        self.req = SetThreshold.Request()

    def send_request(self,threshold):
        self.req.threshold = threshold
        self.future = self.client.call_async(self.req)
        # “请发送这个服务请求，但不要卡在这里等它回来”。call_async 会立即返回一个 Future对象（一个“未来的承诺”），
        # 然后程序继续往下执行
        self.get_logger().info(f'发送请求：设置阈值为{threshold}°C')

def main(args=None):
    rclpy.init(args=args)
    client_node = ThresholdClient()
    
    try:
        threshold_value = int(sys.argv[1])
    except (IndexError,ValueError):
        threshold_value = 15

    client_node.send_request(threshold_value)

    # 等待服务响应
    while rclpy.ok():
        rclpy.spin_once(client_node)
        if client_node.future.done():
            try:
                response = client_node.future.result()
                if response.success:
                    client_node.get_logger().info('服务调用成功：阈值已设置')
                else:
                    client_node.get_logger().error('服务调用失败')
            except Exception as e:
                client_node.get_logger().error(f'服务调用异常:{e}')
            break

    client_node.destroy_node()
    rclpy.shutdown()

File: practice/weather_ws/src/weather_monitor_cpp/src/client_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "interfaces/srv/set_threshold.hpp"

#include <chrono>
#include <memory>

using namespace std::chrono_literals;

int main(int argc,char **argv) {
    rclcpp::init(argc,argv);

    // 我们只需要一个最小化的节点来进行客户端调用
    auto node = rclcpp::Node::make_shared("threshold_client_cpp");

    // 1.创建客户端
    auto client = node->create_client<interfaces::srv::SetThreshold>("set_weather_threshold");

    // 2.等待服务上线
    while(!client->wait_for_service(1s)){
        if(!rclcpp::ok()){
            RCLCPP_ERROR(node->get_logger(),"客户端中断，退出...");
            return 1;
        }
        RCLCPP_INFO(node->get_logger(),"服务不可用，正在等待...");
    }

    // 3.构造请求
    auto request = std::make_shared<interfaces::srv::SetThreshold::Request>();
    request->threshold = 10; // 我们将新阈值硬编码为 10

    // 4.异步发送并等待
    auto future = client->async_send_request(request);

    RCLCPP_INFO(node->get_logger(),"发送请求：设置阈值为10");

    if(rclcpp::spin_until_future_complete(node,future)== rclcpp::FutureReturnCode::SUCCESS){
        auto response = future.get();
        if(response->success){
            RCLCPP_INFO(node->get_logger(),"服务调用成功：阈值已设置");
        }
        else{
            RCLCPP_INFO(node->get_logger(),"服务调用失败");
        } 
    }
    else{
        RCLCPP_ERROR(node->get_logger(),"服务调用异常");
    }

    rclcpp::shutdown();
    return 0;

}
